contracts/libraries/PositionValue.sol

```diff
@@ -0,0 +1,226 @@
+    /// @param positionManager The Uniswap V3 NonfungiblePositionManager
+    /// @param tokenId The tokenId of the token for which to get the total fees owed
+    /// @return amount0 The amount of fees owed in token0
+    /// @return amount1 The amount of fees owed in token1
+    function fees(INonfungiblePositionManager positionManager, uint256 tokenId)
+        internal
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (
+            ,
+            ,
+            address token0,
+            address token1,
+            uint24 fee,
+            int24 tickLower,
+            int24 tickUpper,
+            uint128 liquidity,
+            uint256 positionFeeGrowthInside0LastX128,
+            uint256 positionFeeGrowthInside1LastX128,
+            uint256 tokensOwed0,
+            uint256 tokensOwed1
+        ) = positionManager.positions(tokenId);
+
+        return
+            _fees(
+                positionManager,
+                FeeParams({
+                    token0: token0,
+                    token1: token1,
+                    fee: fee,
+                    tickLower: tickLower,
+                    tickUpper: tickUpper,
+                    liquidity: liquidity,
+                    positionFeeGrowthInside0LastX128: positionFeeGrowthInside0LastX128,
+                    positionFeeGrowthInside1LastX128: positionFeeGrowthInside1LastX128,
+                    tokensOwed0: tokensOwed0,
+                    tokensOwed1: tokensOwed1
+                })
+            );
+    }
+
+    function _fees(INonfungiblePositionManager positionManager, FeeParams memory feeParams)
+        private
+        view
+        returns (uint256 amount0, uint256 amount1)
+    {
+        (uint256 poolFeeGrowthInside0LastX128, uint256 poolFeeGrowthInside1LastX128) =
+            _getFeeGrowthInside(
+                positionManager,
+                feeParams.token0,
+                feeParams.token1,
+                feeParams.fee,
+                feeParams.tickLower,
+                feeParams.tickUpper
+            );
+
+        amount0 =
+            FullMath.mulDiv(
+                poolFeeGrowthInside0LastX128 - feeParams.positionFeeGrowthInside0LastX128,
+                feeParams.liquidity,
+                FixedPoint128.Q128
+            ) +
+            feeParams.tokensOwed0;
+
+        amount1 =
+            FullMath.mulDiv(
+                poolFeeGrowthInside1LastX128 - feeParams.positionFeeGrowthInside1LastX128,
+                feeParams.liquidity,
+                FixedPoint128.Q128
+            ) +
+            feeParams.tokensOwed1;
+    }
+
+    function _getFeeGrowthInside(
+        INonfungiblePositionManager positionManager,
+        address token0,
+        address token1,
+        uint24 fee,
+        int24 tickLower,
+        int24 tickUpper
+    ) private view returns (uint256 feeGrowthInside0X128, uint256 feeGrowthInside1X128) {
+        IUniswapV3Pool pool =
+            IUniswapV3Pool(
```

Comment:
seems like we should just be passing around pool in the fee params, rather than token0/token1/fee

Reply:
I already had them defined though. You think I should sload them again from the pool?

Reply:
i mean that we only use token0/token1/fee to compute the pool, but we are passing around 3 variables instead of the 1 pool address, so you can compute pool in the caller and pass it in to _getFeeGrowthInside instead of all 4 of 
````
        INonfungiblePositionManager positionManager,
        address token0,
        address token1,
        uint24 fee,
```

Reply:
ohhhhh